// core-types-graphql/dist/lib/gql-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isRequired = exports.gqlStripRequired = exports.gqlMaybeRequiredNode = exports.gqlListTypeNode = exports.gqlNamedTypeNode = exports.gqlUnionType = exports.gqlNameNode = void 0;
const graphql_1 = require("graphql");
function gqlNameNode(name) {
    return { kind: graphql_1.Kind.NAME, value: name };
}
exports.gqlNameNode = gqlNameNode;
function gqlUnionType({ name, description, types }) {
    return {
        kind: graphql_1.Kind.UNION_TYPE_DEFINITION,
        name,
        ...(description ? { description } : {}),
        types,
    };
}
exports.gqlUnionType = gqlUnionType;
function gqlNamedTypeNode(name) {
    return { kind: graphql_1.Kind.NAMED_TYPE, name: gqlNameNode(name) };
}
exports.gqlNamedTypeNode = gqlNamedTypeNode;
function gqlListTypeNode(type) {
    return { kind: graphql_1.Kind.LIST_TYPE, type };
}
exports.gqlListTypeNode = gqlListTypeNode;
function gqlMaybeRequiredNode(type, required) {
    return !required ? type : { kind: graphql_1.Kind.NON_NULL_TYPE, type };
}
exports.gqlMaybeRequiredNode = gqlMaybeRequiredNode;
function gqlStripRequired(type) {
    return isRequired(type) ? type.type : type;
}
exports.gqlStripRequired = gqlStripRequired;
function isRequired(type) {
    return type.kind === graphql_1.Kind.NON_NULL_TYPE;
}
exports.isRequired = isRequired;
