// typeconv/dist/convert-suretype.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSureTypeWriter = exports.getSureTypeReader = void 0;
const core_types_suretype_1 = require("core-types-suretype");
const package_1 = require("./package");
const format_graph_1 = require("./format-graph");
const utils_1 = require("./utils");
function getSureTypeReader(options) {
    return {
        kind: 'st',
        managedRead: true,
        read(filename, { warn }) {
            return (0, core_types_suretype_1.convertSureTypeToCoreTypes)(filename, {
                warn,
                userPackage: package_1.userPackage,
                userPackageUrl: package_1.userPackageUrl,
                ...options,
            });
        },
        shortcut: {
            async jsc(filename, { warn }) {
                const result = await (0, core_types_suretype_1.convertSuretypeToJsonSchema)(filename, {
                    warn,
                    userPackage: package_1.userPackage,
                    userPackageUrl: package_1.userPackageUrl,
                    ...options,
                });
                const { convertedTypes, notConvertedTypes, data } = result;
                return {
                    convertedTypes, notConvertedTypes, data: (0, utils_1.stringify)(data),
                };
            },
        },
    };
}
exports.getSureTypeReader = getSureTypeReader;
function getSureTypeWriter(options) {
    return {
        kind: 'st',
        write(doc, { warn, filename, sourceFilename }) {
            return (0, core_types_suretype_1.convertCoreTypesToSureType)(doc, {
                warn,
                filename,
                sourceFilename,
                userPackage: package_1.userPackage,
                userPackageUrl: package_1.userPackageUrl,
                ...options,
            });
        },
        shortcut: {
            jsc(dataString, { warn, filename, sourceFilename }) {
                return (0, core_types_suretype_1.convertJsonSchemaToSureType)(dataString, {
                    warn,
                    filename,
                    sourceFilename,
                    userPackage: package_1.userPackage,
                    userPackageUrl: package_1.userPackageUrl,
                    ...options
                });
            },
        },
    };
}
exports.getSureTypeWriter = getSureTypeWriter;
(0, format_graph_1.registerReader)(getSureTypeReader());
(0, format_graph_1.registerWriter)(getSureTypeWriter());
