// typeconv/dist/convert-core-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCoreTypesWriter = exports.getCoreTypesReader = void 0;
const utils_1 = require("./utils");
const format_graph_1 = require("./format-graph");
function getCoreTypesReader() {
    return {
        kind: 'ct',
        read(source) {
            const doc = JSON.parse(source);
            return {
                data: doc,
                convertedTypes: doc.types.map(({ name }) => name),
                notConvertedTypes: [],
            };
        },
    };
}
exports.getCoreTypesReader = getCoreTypesReader;
function getCoreTypesWriter() {
    return {
        kind: 'ct',
        write(doc) {
            return {
                data: (0, utils_1.stringify)(doc),
                convertedTypes: doc.types.map(({ name }) => name),
                notConvertedTypes: [],
            };
        },
    };
}
exports.getCoreTypesWriter = getCoreTypesWriter;
(0, format_graph_1.registerReader)(getCoreTypesReader());
(0, format_graph_1.registerWriter)(getCoreTypesWriter());
