// typeconv/dist/convert-typescript.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTypeScriptWriter = exports.getTypeScriptReader = void 0;
const core_types_ts_1 = require("core-types-ts");
const package_1 = require("./package");
const format_graph_1 = require("./format-graph");
function getTypeScriptReader(tsOptions) {
    return {
        kind: 'ts',
        read(source, { warn, filename }) {
            return (0, core_types_ts_1.convertTypeScriptToCoreTypes)(source, {
                warn,
                ...tsOptions,
            });
        }
    };
}
exports.getTypeScriptReader = getTypeScriptReader;
function getTypeScriptWriter(tsOptions) {
    return {
        kind: 'ts',
        write(doc, { warn, filename, sourceFilename }) {
            return (0, core_types_ts_1.convertCoreTypesToTypeScript)(doc, {
                warn,
                filename,
                sourceFilename,
                userPackage: package_1.userPackage,
                userPackageUrl: package_1.userPackageUrl,
                ...tsOptions,
            });
        }
    };
}
exports.getTypeScriptWriter = getTypeScriptWriter;
(0, format_graph_1.registerReader)(getTypeScriptReader());
(0, format_graph_1.registerWriter)(getTypeScriptWriter());
