// typeconv/dist/convert-graphql.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getGraphQLWriter = exports.getGraphQLReader = void 0;
const core_types_graphql_1 = require("core-types-graphql");
const package_1 = require("./package");
const format_graph_1 = require("./format-graph");
function getGraphQLReader(graphqlOptions) {
    return {
        kind: 'gql',
        read(source, options) {
            return (0, core_types_graphql_1.convertGraphqlToCoreTypes)(source, {
                ...options,
                ...graphqlOptions,
            });
        }
    };
}
exports.getGraphQLReader = getGraphQLReader;
function getGraphQLWriter(graphqlOptions) {
    return {
        kind: 'gql',
        write(doc, options) {
            return (0, core_types_graphql_1.convertCoreTypesToGraphql)(doc, {
                warn: options.warn,
                filename: options.filename,
                sourceFilename: options.sourceFilename,
                userPackage: package_1.userPackage,
                userPackageUrl: package_1.userPackageUrl,
                ...graphqlOptions,
            });
        }
    };
}
exports.getGraphQLWriter = getGraphQLWriter;
(0, format_graph_1.registerReader)(getGraphQLReader());
(0, format_graph_1.registerWriter)(getGraphQLWriter());
