// suretype/dist/annotations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNames = exports.getName = exports.getAnnotations = exports.annotateValidator = exports.AnnotationsHolder = void 0;
const validator_1 = require("./validators/core/validator");
const validator_2 = require("./validators/raw/validator");
class AnnotationsHolder {
    constructor(options) {
        this.options = options;
    }
}
exports.AnnotationsHolder = AnnotationsHolder;
function annotateValidator(validator, annotations) {
    validator_1.exposeCoreValidator(validator)._annotations = annotations;
    return validator;
}
exports.annotateValidator = annotateValidator;
function getAnnotations(validator) {
    var _a;
    const annotations = (_a = validator_1.exposeCoreValidator(validator)._annotations) === null || _a === void 0 ? void 0 : _a.options;
    const raw = validator_2.getRaw(validator);
    if (raw && raw.fragment) {
        if (!(annotations === null || annotations === void 0 ? void 0 : annotations.name))
            return { ...annotations, name: raw.fragment };
    }
    return annotations;
}
exports.getAnnotations = getAnnotations;
function getName(validator) {
    var _a, _b;
    const name = (_b = (_a = validator_1.exposeCoreValidator(validator)._annotations) === null || _a === void 0 ? void 0 : _a.options) === null || _b === void 0 ? void 0 : _b.name;
    const raw = validator_2.getRaw(validator);
    if (!name && raw && raw.fragment) {
        return raw.fragment;
    }
    return name;
}
exports.getName = getName;
function getNames(validator) {
    var _a, _b;
    const name = (_b = (_a = validator_1.exposeCoreValidator(validator)._annotations) === null || _a === void 0 ? void 0 : _a.options) === null || _b === void 0 ? void 0 : _b.name;
    const raw = validator_2.getRaw(validator);
    const otherNames = raw && raw.fragment
        ? Object.keys(raw.toSchema().definitions)
        : [];
    return name ? [...new Set([name, ...otherNames])] : otherNames;
}
exports.getNames = getNames;
