// suretype/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAnnotations = exports.getValidatorSchema = exports.extractSingleJsonSchema = exports.extractJsonSchema = exports.validate = exports.compile = exports.ValidationError = void 0;
__exportStar(require("./api"), exports);
__exportStar(require("./errors"), exports);
var validation_error_1 = require("./validation-error");
Object.defineProperty(exports, "ValidationError", { enumerable: true, get: function () { return validation_error_1.ValidationError; } });
__exportStar(require("./types"), exports);
const json_schema_1 = require("./json-schema");
Object.defineProperty(exports, "compile", { enumerable: true, get: function () { return json_schema_1.compile; } });
Object.defineProperty(exports, "validate", { enumerable: true, get: function () { return json_schema_1.validate; } });
var extract_json_schema_1 = require("./extract-json-schema");
Object.defineProperty(exports, "extractJsonSchema", { enumerable: true, get: function () { return extract_json_schema_1.extractJsonSchema; } });
Object.defineProperty(exports, "extractSingleJsonSchema", { enumerable: true, get: function () { return extract_json_schema_1.extractSingleJsonSchema; } });
var validation_1 = require("./validation");
Object.defineProperty(exports, "getValidatorSchema", { enumerable: true, get: function () { return validation_1.getValidatorSchema; } });
var annotations_1 = require("./annotations");
Object.defineProperty(exports, "getAnnotations", { enumerable: true, get: function () { return annotations_1.getAnnotations; } });
