// awesome-ajv-errors/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.suggestAnotherType = exports.isSimpleValue = exports.getTypedValueKey = exports.getTypedValue = exports.getValueType = exports.getTypedContext = exports.getError = void 0;
function getError(context) {
    return context.error;
}
exports.getError = getError;
function getTypedContext(context) {
    return context;
}
exports.getTypedContext = getTypedContext;
const getValueType = (value) => value === null ? 'null' : typeof value;
exports.getValueType = getValueType;
const getTypedValue = (value) => value == null || typeof value !== 'object'
    ? { value, type: exports.getValueType(value), isSimple: true }
    : Array.isArray(value)
        ? { value: '[...]', type: 'array', isSimple: false }
        : { value: '{...}', type: 'object', isSimple: false };
exports.getTypedValue = getTypedValue;
const getTypedValueKey = ({ type, value }) => type + "::" + value;
exports.getTypedValueKey = getTypedValueKey;
const isSimpleValue = (possible) => possible === null ||
    typeof possible === 'boolean' ||
    typeof possible === 'number' ||
    typeof possible === 'string';
exports.isSimpleValue = isSimpleValue;
function suggestAnotherType(value, toType) {
    const { isSimple, type: fromType } = exports.getTypedValue(value);
    if (!isSimple)
        return;
    const convert = () => {
        if (fromType === 'null') {
            if (toType === 'string')
                return 'null';
            else if (toType === 'number')
                return 0;
            else if (toType === 'boolean')
                return false;
        }
        else if (fromType === 'boolean') {
            if (toType === 'string')
                return `${value}`;
            else if (toType === 'number')
                return value ? 1 : 0;
            else if (toType === 'null')
                return value ? undefined : null;
        }
        else if (fromType === 'number') {
            if (toType === 'string')
                return `${value}`;
            else if (toType === 'boolean')
                return value === 0 ? false : value === 1 ? true : undefined;
            else if (toType === 'null')
                return value === 0 ? null : undefined;
        }
        else if (fromType === 'string') {
            if (toType === 'number')
                return isNaN(Number(value)) ? undefined : Number(value);
            else if (toType === 'boolean')
                return value === "true"
                    ? true
                    : value === "false"
                        ? false
                        : undefined;
            else if (toType === 'null')
                return value === "null" ? null : undefined;
        }
        return;
    };
    const ret = convert();
    if (ret === undefined)
        return undefined;
    return exports.getTypedValue(ret);
}
exports.suggestAnotherType = suggestAnotherType;
